{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook 2: finding the transcript lengths\n",
    "\n",
    "author: jake siegel\n",
    "\n",
    "date: 11.29.2017\n",
    "\n",
    "The purpose of this notebook is to figure out how to calculate the transcript length for the RNA Seq data using the .gtf file downloaded for gencode v17 from gencode.  The file is in /data/raw\n",
    "\n",
    "The gtf is massive... it's 1.03GB.  This may be too big to read in all at once.  It might have to be read in in chunks.\n",
    "\n",
    "The v17.annotation file contains fewer genes than the RNA-Seq processed dataset (~57,400)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2823730, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        1           2      3      4  \\\n",
       "0  HAVANA        gene  11869  14412   \n",
       "1  HAVANA  transcript  11869  14409   \n",
       "2  HAVANA        exon  11869  12227   \n",
       "3  HAVANA        exon  12613  12721   \n",
       "4  HAVANA        exon  13221  14409   \n",
       "\n",
       "                                                   8  \n",
       "0  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  \n",
       "1  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  \n",
       "2  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  \n",
       "3  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  \n",
       "4  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import time\n",
    "\n",
    "# data_src = '../data/raw/gencode.v17.annotation.gtf'\n",
    "data_src = '../data/raw/gencode.v17.chr_patch_hapl_scaff.annotation.gtf'\n",
    "df = pd.read_csv(data_src, skiprows=5, header=None, usecols=[1,2,3,4,8],sep = '\\t')\n",
    "print df.shape\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many genes are there in this build?\n",
    "df[8].astype(str).str.split('\"', expand=True)[1].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Can I just use the Ensembl genes? No.\n",
    "test = df.loc[df[1]==\"ENSEMBL\"]\n",
    "test[8].astype(str).str.split('\"', expand=True)[1].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>8</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        1           2      3      4  \\\n",
       "0  HAVANA        gene  11869  14412   \n",
       "1  HAVANA  transcript  11869  14409   \n",
       "2  HAVANA        exon  11869  12227   \n",
       "3  HAVANA        exon  12613  12721   \n",
       "4  HAVANA        exon  13221  14409   \n",
       "\n",
       "                                                   8            gene_id  \\\n",
       "0  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "1  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "2  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "3  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "4  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "\n",
       "       transcript_id  \n",
       "0  ENSG00000223972.4  \n",
       "1  ENST00000456328.2  \n",
       "2  ENST00000456328.2  \n",
       "3  ENST00000456328.2  \n",
       "4  ENST00000456328.2  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['gene_id']=df[8].astype(str).str.split('\"', expand=True)[1]\n",
    "df['transcript_id']=df[8].astype(str).str.split('\"', expand=True)[3]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>8</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>2544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>1189</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        1           2      3      4  \\\n",
       "0  HAVANA        gene  11869  14412   \n",
       "1  HAVANA  transcript  11869  14409   \n",
       "2  HAVANA        exon  11869  12227   \n",
       "3  HAVANA        exon  12613  12721   \n",
       "4  HAVANA        exon  13221  14409   \n",
       "\n",
       "                                                   8            gene_id  \\\n",
       "0  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "1  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "2  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "3  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "4  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  ENSG00000223972.4   \n",
       "\n",
       "       transcript_id    bp  \n",
       "0  ENSG00000223972.4  2544  \n",
       "1  ENST00000456328.2  2541  \n",
       "2  ENST00000456328.2   359  \n",
       "3  ENST00000456328.2   109  \n",
       "4  ENST00000456328.2  1189  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['bp']=df[4]-df[3]+1\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OK, so here is where we need to figure out how to calculate the \"right\" transcript length for each transcript.  \n",
    "\n",
    "The library prep protocol in the paper is a poly-A-selective protocol, so we need the length of the mature transcript.  That means summing over exons.  Many of the genes in this annotation have multiple entries (for example, see below).  \n",
    "\n",
    "So, my plan is to sum over exons for each transcript_id and then take the maximum (by length) of all of the mature transcript_id's for a given gene_id as the length of that gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.iloc[:20,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of gene_id: (62798,)\n",
      "number of transcript_id: (211826,)\n"
     ]
    }
   ],
   "source": [
    "# There is a row that is annotated 'gene' in column 2 that is unique for each gene_id.\n",
    "# That makes for a convenient hierarchy for keeping track of the calculated transcript lengths.\n",
    "# I am going to put the calculted mature transcript length for each transcript in the row annoted as \"transcript\"\n",
    "# in column 2.\n",
    "# Then I am going to put the maximum calculated transcript length for a given gene in the row annoted \"gene\".\n",
    "print 'number of gene_id:', df['gene_id'].loc[df[2]=='gene'].unique().shape\n",
    "print 'number of transcript_id:', df['transcript_id'].loc[df[2]=='transcript'].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "summing over 1 transcript id took: 3.26815080643 seconds\n"
     ]
    }
   ],
   "source": [
    "# Here is a way to do the summation for a given transcript:\n",
    "import time\n",
    "\n",
    "start = time.time()\n",
    "temp = df['bp'].loc[(df['transcript_id']=='ENST00000515242.2')&(df[2]=='exon')].sum()\n",
    "end = time.time()\n",
    "print 'summing over 1 transcript id took:', (end-start),'seconds'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yeah, that's going to take 8 days to do it that way.  Need to think of a different option."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "groupby summation for 4 transcipt ids takes: 0.00754117965698 seconds\n"
     ]
    }
   ],
   "source": [
    "# Here we loop over all unique transcripts and put the calculated length in the row annotated \"transcript\" in column 2.\n",
    "df2 = pd.DataFrame()\n",
    "df2 = df.iloc[:20,:]\n",
    "df2.shape\n",
    "start = time.time()\n",
    "df2.groupby(['gene_id','transcript_id',2])['bp'].sum()\n",
    "end = time.time()\n",
    "print 'groupby summation for 4 transcipt ids takes:',(end-start),'seconds'\n",
    "\n",
    "\n",
    "# for transcript in df['transcript_id'].loc[df[2]!='gene'].unique():\n",
    "#     df['bp'].loc[(df['transcript_id']==transcript)&(df[2]=='transcript')]=df['bp'].loc[(df['transcript_id']==transcript)&(df[2]=='exon')].sum()\n",
    "\n",
    "# df.iloc[:20,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>2</th>\n",
       "      <th>bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.10</td>\n",
       "      <td>ENSG00000000003.10</td>\n",
       "      <td>gene</td>\n",
       "      <td>11322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000003.10</td>\n",
       "      <td>ENST00000373020.4</td>\n",
       "      <td>CDS</td>\n",
       "      <td>735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003.10</td>\n",
       "      <td>ENST00000373020.4</td>\n",
       "      <td>UTR</td>\n",
       "      <td>1471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000003.10</td>\n",
       "      <td>ENST00000373020.4</td>\n",
       "      <td>exon</td>\n",
       "      <td>2206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000003.10</td>\n",
       "      <td>ENST00000373020.4</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id       transcript_id            2     bp\n",
       "0  ENSG00000000003.10  ENSG00000000003.10         gene  11322\n",
       "1  ENSG00000000003.10   ENST00000373020.4          CDS    735\n",
       "2  ENSG00000000003.10   ENST00000373020.4          UTR   1471\n",
       "3  ENSG00000000003.10   ENST00000373020.4         exon   2206\n",
       "4  ENSG00000000003.10   ENST00000373020.4  start_codon      3"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lengths = pd.DataFrame()\n",
    "lengths = df.groupby(['gene_id','transcript_id',2])['bp'].sum()\n",
    "lengths = lengths.reset_index()\n",
    "lengths.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we just need to select the maximum exon length for each gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>bp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.10</td>\n",
       "      <td>2206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>1339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.8</td>\n",
       "      <td>1161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.8</td>\n",
       "      <td>3477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.12</td>\n",
       "      <td>4355</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id    bp\n",
       "0  ENSG00000000003.10  2206\n",
       "1   ENSG00000000005.5  1339\n",
       "2   ENSG00000000419.8  1161\n",
       "3   ENSG00000000457.8  3477\n",
       "4  ENSG00000000460.12  4355"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_lengths = lengths.loc[lengths[2]=='exon'].groupby('gene_id')['bp'].max().reset_index()\n",
    "transcript_lengths.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(62798, 2)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_lengths.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_lengths.to_csv('../data/interim/transcipt_lengths.csv', encoding='utf-8', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final Remarks\n",
    "\n",
    "It seems this did the trick.  This version of the GENCODE v17 annotation has gene_ids that match the read count data and is contains more genes than the read count file, so hopefully every read gene has a transcript length.\n",
    "\n",
    "The processed lengths are in the /data/interim folder.\n",
    "\n",
    "The next step is to go back to the read count data and normalize it to TPM. I probably need to figure out what those spike-in control reads mean.  Bare minimum, I should dump them.  I wonder, though, if people use them as a way to normalize the reads.  That seems silly, but there might be some value in that data that I am overlooking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2823730, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1           2      3      4  5  6  7  \\\n",
       "0  chr1  HAVANA        gene  11869  14412  .  +  .   \n",
       "1  chr1  HAVANA  transcript  11869  14409  .  +  .   \n",
       "2  chr1  HAVANA        exon  11869  12227  .  +  .   \n",
       "3  chr1  HAVANA        exon  12613  12721  .  +  .   \n",
       "4  chr1  HAVANA        exon  13221  14409  .  +  .   \n",
       "\n",
       "                                                   8  \n",
       "0  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  \n",
       "1  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  \n",
       "2  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  \n",
       "3  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  \n",
       "4  gene_id \"ENSG00000223972.4\"; transcript_id \"EN...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import time\n",
    "\n",
    "# data_src = '../data/raw/gencode.v17.annotation.gtf'\n",
    "data_src = '../data/raw/gencode.v17.chr_patch_hapl_scaff.annotation.gtf'\n",
    "df = pd.read_csv(data_src, skiprows=5, header=None,sep = '\\t')\n",
    "print df.shape\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>25</th>\n",
       "      <th>26</th>\n",
       "      <th>27</th>\n",
       "      <th>28</th>\n",
       "      <th>29</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "      <th>34</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000515242.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000515242.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000515242.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000515242.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000518655.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000518655.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000518655.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000518655.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000518655.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000438504.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000438504.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000438504.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000438504.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000438504.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000438504.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000438504.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000438504.2</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823700</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823701</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823702</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823703</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823704</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823705</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823706</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823707</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823708</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823709</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823710</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823711</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823712</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374457.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823713</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823714</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823715</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823716</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823717</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823718</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823719</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823720</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823721</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823722</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823723</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823724</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823725</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823726</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823727</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823728</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2823729</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>ENSG00000215699.3</td>\n",
       "      <td>; transcript_id</td>\n",
       "      <td>ENST00000374449.3</td>\n",
       "      <td>; gene_type</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>; gene_status</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>; gene_name</td>\n",
       "      <td>SRSF10</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2823730 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0                  1                 2                  3   \\\n",
       "0        gene_id   ENSG00000223972.4  ; transcript_id   ENSG00000223972.4   \n",
       "1        gene_id   ENSG00000223972.4  ; transcript_id   ENST00000456328.2   \n",
       "2        gene_id   ENSG00000223972.4  ; transcript_id   ENST00000456328.2   \n",
       "3        gene_id   ENSG00000223972.4  ; transcript_id   ENST00000456328.2   \n",
       "4        gene_id   ENSG00000223972.4  ; transcript_id   ENST00000456328.2   \n",
       "5        gene_id   ENSG00000223972.4  ; transcript_id   ENST00000515242.2   \n",
       "6        gene_id   ENSG00000223972.4  ; transcript_id   ENST00000515242.2   \n",
       "7        gene_id   ENSG00000223972.4  ; transcript_id   ENST00000515242.2   \n",
       "8        gene_id   ENSG00000223972.4  ; transcript_id   ENST00000515242.2   \n",
       "9        gene_id   ENSG00000223972.4  ; transcript_id   ENST00000518655.2   \n",
       "10       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000518655.2   \n",
       "11       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000518655.2   \n",
       "12       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000518655.2   \n",
       "13       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000518655.2   \n",
       "14       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000450305.2   \n",
       "15       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000450305.2   \n",
       "16       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000450305.2   \n",
       "17       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000450305.2   \n",
       "18       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000450305.2   \n",
       "19       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000450305.2   \n",
       "20       gene_id   ENSG00000223972.4  ; transcript_id   ENST00000450305.2   \n",
       "21       gene_id   ENSG00000227232.4  ; transcript_id   ENSG00000227232.4   \n",
       "22       gene_id   ENSG00000227232.4  ; transcript_id   ENST00000438504.2   \n",
       "23       gene_id   ENSG00000227232.4  ; transcript_id   ENST00000438504.2   \n",
       "24       gene_id   ENSG00000227232.4  ; transcript_id   ENST00000438504.2   \n",
       "25       gene_id   ENSG00000227232.4  ; transcript_id   ENST00000438504.2   \n",
       "26       gene_id   ENSG00000227232.4  ; transcript_id   ENST00000438504.2   \n",
       "27       gene_id   ENSG00000227232.4  ; transcript_id   ENST00000438504.2   \n",
       "28       gene_id   ENSG00000227232.4  ; transcript_id   ENST00000438504.2   \n",
       "29       gene_id   ENSG00000227232.4  ; transcript_id   ENST00000438504.2   \n",
       "...           ...                ...               ...                ...   \n",
       "2823700  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823701  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823702  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823703  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823704  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823705  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823706  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823707  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823708  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823709  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823710  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823711  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823712  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374457.3   \n",
       "2823713  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823714  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823715  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823716  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823717  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823718  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823719  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823720  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823721  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823722  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823723  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823724  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823725  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823726  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823727  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823728  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "2823729  gene_id   ENSG00000215699.3  ; transcript_id   ENST00000374449.3   \n",
       "\n",
       "                   4               5               6      7             8   \\\n",
       "0        ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "1        ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "2        ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "3        ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "4        ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "5        ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "6        ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "7        ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "8        ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "9        ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "10       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "11       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "12       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "13       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "14       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "15       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "16       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "17       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "18       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "19       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "20       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "21       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "22       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "23       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "24       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "25       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "26       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "27       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "28       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "29       ; gene_type       pseudogene  ; gene_status   KNOWN  ; gene_name    \n",
       "...               ...             ...             ...    ...           ...   \n",
       "2823700  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823701  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823702  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823703  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823704  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823705  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823706  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823707  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823708  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823709  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823710  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823711  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823712  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823713  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823714  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823715  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823716  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823717  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823718  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823719  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823720  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823721  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823722  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823723  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823724  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823725  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823726  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823727  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823728  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "2823729  ; gene_type   protein_coding  ; gene_status   KNOWN  ; gene_name    \n",
       "\n",
       "              9   ...     25    26    27    28    29    30    31    32    33  \\\n",
       "0        DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "1        DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "2        DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "3        DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "4        DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "5        DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "6        DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "7        DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "8        DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "9        DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "10       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "11       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "12       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "13       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "14       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "15       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "16       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "17       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "18       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "19       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "20       DDX11L1  ...   None  None  None  None  None  None  None  None  None   \n",
       "21        WASH7P  ...   None  None  None  None  None  None  None  None  None   \n",
       "22        WASH7P  ...   None  None  None  None  None  None  None  None  None   \n",
       "23        WASH7P  ...   None  None  None  None  None  None  None  None  None   \n",
       "24        WASH7P  ...   None  None  None  None  None  None  None  None  None   \n",
       "25        WASH7P  ...   None  None  None  None  None  None  None  None  None   \n",
       "26        WASH7P  ...   None  None  None  None  None  None  None  None  None   \n",
       "27        WASH7P  ...   None  None  None  None  None  None  None  None  None   \n",
       "28        WASH7P  ...   None  None  None  None  None  None  None  None  None   \n",
       "29        WASH7P  ...   None  None  None  None  None  None  None  None  None   \n",
       "...          ...  ...    ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "2823700   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823701   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823702   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823703   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823704   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823705   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823706   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823707   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823708   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823709   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823710   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823711   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823712   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823713   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823714   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823715   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823716   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823717   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823718   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823719   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823720   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823721   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823722   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823723   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823724   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823725   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823726   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823727   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823728   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "2823729   SRSF10  ...   None  None  None  None  None  None  None  None  None   \n",
       "\n",
       "           34  \n",
       "0        None  \n",
       "1        None  \n",
       "2        None  \n",
       "3        None  \n",
       "4        None  \n",
       "5        None  \n",
       "6        None  \n",
       "7        None  \n",
       "8        None  \n",
       "9        None  \n",
       "10       None  \n",
       "11       None  \n",
       "12       None  \n",
       "13       None  \n",
       "14       None  \n",
       "15       None  \n",
       "16       None  \n",
       "17       None  \n",
       "18       None  \n",
       "19       None  \n",
       "20       None  \n",
       "21       None  \n",
       "22       None  \n",
       "23       None  \n",
       "24       None  \n",
       "25       None  \n",
       "26       None  \n",
       "27       None  \n",
       "28       None  \n",
       "29       None  \n",
       "...       ...  \n",
       "2823700  None  \n",
       "2823701  None  \n",
       "2823702  None  \n",
       "2823703  None  \n",
       "2823704  None  \n",
       "2823705  None  \n",
       "2823706  None  \n",
       "2823707  None  \n",
       "2823708  None  \n",
       "2823709  None  \n",
       "2823710  None  \n",
       "2823711  None  \n",
       "2823712  None  \n",
       "2823713  None  \n",
       "2823714  None  \n",
       "2823715  None  \n",
       "2823716  None  \n",
       "2823717  None  \n",
       "2823718  None  \n",
       "2823719  None  \n",
       "2823720  None  \n",
       "2823721  None  \n",
       "2823722  None  \n",
       "2823723  None  \n",
       "2823724  None  \n",
       "2823725  None  \n",
       "2823726  None  \n",
       "2823727  None  \n",
       "2823728  None  \n",
       "2823729  None  \n",
       "\n",
       "[2823730 rows x 35 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[8].astype(str).str.split('\"', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
