{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook 1: loading & processing data\n",
    "author: jake siegel\n",
    "\n",
    "date: 11.28.2017\n",
    "\n",
    "This notebook loads the asynchronous single cell RNA Seq data from the Karlson et al. 2017 JMB paper.  It's in /references.\n",
    "\n",
    "Here is the reference:\n",
    "\n",
    "https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-6142/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(59838, 97)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>S1</th>\n",
       "      <th>S2</th>\n",
       "      <th>S3</th>\n",
       "      <th>S4</th>\n",
       "      <th>S5</th>\n",
       "      <th>S6</th>\n",
       "      <th>S7</th>\n",
       "      <th>S8</th>\n",
       "      <th>S9</th>\n",
       "      <th>...</th>\n",
       "      <th>S87</th>\n",
       "      <th>S88</th>\n",
       "      <th>S89</th>\n",
       "      <th>S90</th>\n",
       "      <th>S91</th>\n",
       "      <th>S92</th>\n",
       "      <th>S93</th>\n",
       "      <th>S94</th>\n",
       "      <th>S95</th>\n",
       "      <th>S96</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.10</td>\n",
       "      <td>360</td>\n",
       "      <td>5</td>\n",
       "      <td>437</td>\n",
       "      <td>136</td>\n",
       "      <td>328</td>\n",
       "      <td>253</td>\n",
       "      <td>1101</td>\n",
       "      <td>39</td>\n",
       "      <td>157</td>\n",
       "      <td>...</td>\n",
       "      <td>391</td>\n",
       "      <td>429</td>\n",
       "      <td>148</td>\n",
       "      <td>397</td>\n",
       "      <td>424</td>\n",
       "      <td>317</td>\n",
       "      <td>403</td>\n",
       "      <td>280</td>\n",
       "      <td>470</td>\n",
       "      <td>725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.8</td>\n",
       "      <td>111</td>\n",
       "      <td>421</td>\n",
       "      <td>70</td>\n",
       "      <td>179</td>\n",
       "      <td>93</td>\n",
       "      <td>57</td>\n",
       "      <td>60</td>\n",
       "      <td>35</td>\n",
       "      <td>174</td>\n",
       "      <td>...</td>\n",
       "      <td>63</td>\n",
       "      <td>228</td>\n",
       "      <td>173</td>\n",
       "      <td>115</td>\n",
       "      <td>87</td>\n",
       "      <td>308</td>\n",
       "      <td>92</td>\n",
       "      <td>179</td>\n",
       "      <td>164</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.8</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.12</td>\n",
       "      <td>179</td>\n",
       "      <td>448</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>135</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>151</td>\n",
       "      <td>68</td>\n",
       "      <td>147</td>\n",
       "      <td>84</td>\n",
       "      <td>151</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>169</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Gene_ID   S1   S2   S3   S4   S5   S6    S7  S8   S9 ...   S87  \\\n",
       "0  ENSG00000000003.10  360    5  437  136  328  253  1101  39  157 ...   391   \n",
       "1   ENSG00000000005.5    0    0    0    0    0    0     0   0    0 ...     0   \n",
       "2   ENSG00000000419.8  111  421   70  179   93   57    60  35  174 ...    63   \n",
       "3   ENSG00000000457.8    2    5    0    0    1    1     3   0    5 ...     3   \n",
       "4  ENSG00000000460.12  179  448    0    0  135   47     0   0    0 ...   151   \n",
       "\n",
       "   S88  S89  S90  S91  S92  S93  S94  S95  S96  \n",
       "0  429  148  397  424  317  403  280  470  725  \n",
       "1    0    0    0    0    0    0    0    0    0  \n",
       "2  228  173  115   87  308   92  179  164  104  \n",
       "3    0    2    1    2    0    0    0   38    1  \n",
       "4   68  147   84  151   11    0    0  169   77  \n",
       "\n",
       "[5 rows x 97 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "data_src = '../data/raw/counts.txt'\n",
    "df = pd.read_csv(data_src, sep = '\\t')\n",
    "print df.shape\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(192, 39)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source Name</th>\n",
       "      <th>Comment[ENA_SAMPLE]</th>\n",
       "      <th>Comment[BioSD_SAMPLE]</th>\n",
       "      <th>Characteristics[organism]</th>\n",
       "      <th>Characteristics[cell line]</th>\n",
       "      <th>Characteristics[disease]</th>\n",
       "      <th>Characteristics[organism part]</th>\n",
       "      <th>Characteristics[cell cycle phase]</th>\n",
       "      <th>Material Type</th>\n",
       "      <th>Protocol REF</th>\n",
       "      <th>...</th>\n",
       "      <th>Comment[ENA_RUN]</th>\n",
       "      <th>Comment[FASTQ_URI]</th>\n",
       "      <th>Comment[SPOT_LENGTH]</th>\n",
       "      <th>Comment[READ_INDEX_1_BASE_COORD]</th>\n",
       "      <th>Protocol REF.5</th>\n",
       "      <th>Protocol REF.6</th>\n",
       "      <th>Derived Array Data File</th>\n",
       "      <th>Comment [Derived ArrayExpress FTP file]</th>\n",
       "      <th>Factor Value[cell cycle phase]</th>\n",
       "      <th>Factor Value[single cell identifier]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S73</td>\n",
       "      <td>ERS1979261</td>\n",
       "      <td>SAMEA104354283</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MLS 1765-92</td>\n",
       "      <td>myxoid liposarcoma</td>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>G2/M</td>\n",
       "      <td>cell</td>\n",
       "      <td>P-MTAB-68727</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR2177244</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/004/...</td>\n",
       "      <td>300</td>\n",
       "      <td>151</td>\n",
       "      <td>P-MTAB-68731</td>\n",
       "      <td>P-MTAB-68732</td>\n",
       "      <td>counts.txt</td>\n",
       "      <td>ftp://ftp.ebi.ac.uk/pub/databases/microarray/d...</td>\n",
       "      <td>G2/M</td>\n",
       "      <td>S73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S73</td>\n",
       "      <td>ERS1979261</td>\n",
       "      <td>SAMEA104354283</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MLS 1765-92</td>\n",
       "      <td>myxoid liposarcoma</td>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>G2/M</td>\n",
       "      <td>cell</td>\n",
       "      <td>P-MTAB-68727</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR2177244</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/004/...</td>\n",
       "      <td>74</td>\n",
       "      <td>38</td>\n",
       "      <td>P-MTAB-68731</td>\n",
       "      <td>P-MTAB-68732</td>\n",
       "      <td>counts.txt</td>\n",
       "      <td>ftp://ftp.ebi.ac.uk/pub/databases/microarray/d...</td>\n",
       "      <td>G2/M</td>\n",
       "      <td>S73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S1</td>\n",
       "      <td>ERS1979262</td>\n",
       "      <td>SAMEA104354284</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MLS 1765-92</td>\n",
       "      <td>myxoid liposarcoma</td>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>G1</td>\n",
       "      <td>cell</td>\n",
       "      <td>P-MTAB-68727</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR2177245</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/005/...</td>\n",
       "      <td>302</td>\n",
       "      <td>152</td>\n",
       "      <td>P-MTAB-68731</td>\n",
       "      <td>P-MTAB-68732</td>\n",
       "      <td>counts.txt</td>\n",
       "      <td>ftp://ftp.ebi.ac.uk/pub/databases/microarray/d...</td>\n",
       "      <td>G1</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S1</td>\n",
       "      <td>ERS1979262</td>\n",
       "      <td>SAMEA104354284</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MLS 1765-92</td>\n",
       "      <td>myxoid liposarcoma</td>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>G1</td>\n",
       "      <td>cell</td>\n",
       "      <td>P-MTAB-68727</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR2177245</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/005/...</td>\n",
       "      <td>64</td>\n",
       "      <td>33</td>\n",
       "      <td>P-MTAB-68731</td>\n",
       "      <td>P-MTAB-68732</td>\n",
       "      <td>counts.txt</td>\n",
       "      <td>ftp://ftp.ebi.ac.uk/pub/databases/microarray/d...</td>\n",
       "      <td>G1</td>\n",
       "      <td>S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S9</td>\n",
       "      <td>ERS1979263</td>\n",
       "      <td>SAMEA104354285</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MLS 1765-92</td>\n",
       "      <td>myxoid liposarcoma</td>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>G1</td>\n",
       "      <td>cell</td>\n",
       "      <td>P-MTAB-68727</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR2177246</td>\n",
       "      <td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/006/...</td>\n",
       "      <td>302</td>\n",
       "      <td>152</td>\n",
       "      <td>P-MTAB-68731</td>\n",
       "      <td>P-MTAB-68732</td>\n",
       "      <td>counts.txt</td>\n",
       "      <td>ftp://ftp.ebi.ac.uk/pub/databases/microarray/d...</td>\n",
       "      <td>G1</td>\n",
       "      <td>S9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Source Name Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \\\n",
       "0         S73          ERS1979261        SAMEA104354283   \n",
       "1         S73          ERS1979261        SAMEA104354283   \n",
       "2          S1          ERS1979262        SAMEA104354284   \n",
       "3          S1          ERS1979262        SAMEA104354284   \n",
       "4          S9          ERS1979263        SAMEA104354285   \n",
       "\n",
       "  Characteristics[organism] Characteristics[cell line]  \\\n",
       "0              Homo sapiens                MLS 1765-92   \n",
       "1              Homo sapiens                MLS 1765-92   \n",
       "2              Homo sapiens                MLS 1765-92   \n",
       "3              Homo sapiens                MLS 1765-92   \n",
       "4              Homo sapiens                MLS 1765-92   \n",
       "\n",
       "  Characteristics[disease] Characteristics[organism part]  \\\n",
       "0       myxoid liposarcoma                 adipose tissue   \n",
       "1       myxoid liposarcoma                 adipose tissue   \n",
       "2       myxoid liposarcoma                 adipose tissue   \n",
       "3       myxoid liposarcoma                 adipose tissue   \n",
       "4       myxoid liposarcoma                 adipose tissue   \n",
       "\n",
       "  Characteristics[cell cycle phase] Material Type  Protocol REF  \\\n",
       "0                              G2/M          cell  P-MTAB-68727   \n",
       "1                              G2/M          cell  P-MTAB-68727   \n",
       "2                                G1          cell  P-MTAB-68727   \n",
       "3                                G1          cell  P-MTAB-68727   \n",
       "4                                G1          cell  P-MTAB-68727   \n",
       "\n",
       "                   ...                  Comment[ENA_RUN]  \\\n",
       "0                  ...                        ERR2177244   \n",
       "1                  ...                        ERR2177244   \n",
       "2                  ...                        ERR2177245   \n",
       "3                  ...                        ERR2177245   \n",
       "4                  ...                        ERR2177246   \n",
       "\n",
       "                                  Comment[FASTQ_URI] Comment[SPOT_LENGTH]  \\\n",
       "0  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/004/...                  300   \n",
       "1  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/004/...                   74   \n",
       "2  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/005/...                  302   \n",
       "3  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/005/...                   64   \n",
       "4  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR217/006/...                  302   \n",
       "\n",
       "  Comment[READ_INDEX_1_BASE_COORD] Protocol REF.5 Protocol REF.6  \\\n",
       "0                              151   P-MTAB-68731   P-MTAB-68732   \n",
       "1                               38   P-MTAB-68731   P-MTAB-68732   \n",
       "2                              152   P-MTAB-68731   P-MTAB-68732   \n",
       "3                               33   P-MTAB-68731   P-MTAB-68732   \n",
       "4                              152   P-MTAB-68731   P-MTAB-68732   \n",
       "\n",
       "  Derived Array Data File            Comment [Derived ArrayExpress FTP file]  \\\n",
       "0              counts.txt  ftp://ftp.ebi.ac.uk/pub/databases/microarray/d...   \n",
       "1              counts.txt  ftp://ftp.ebi.ac.uk/pub/databases/microarray/d...   \n",
       "2              counts.txt  ftp://ftp.ebi.ac.uk/pub/databases/microarray/d...   \n",
       "3              counts.txt  ftp://ftp.ebi.ac.uk/pub/databases/microarray/d...   \n",
       "4              counts.txt  ftp://ftp.ebi.ac.uk/pub/databases/microarray/d...   \n",
       "\n",
       "  Factor Value[cell cycle phase]  Factor Value[single cell identifier]  \n",
       "0                           G2/M                                   S73  \n",
       "1                           G2/M                                   S73  \n",
       "2                             G1                                    S1  \n",
       "3                             G1                                    S1  \n",
       "4                             G1                                    S9  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cool.  Now let's load the info file to get the cell cycle stage\n",
    "\n",
    "info_src = '../data/raw/E-MTAB-6142.sdrf.txt'\n",
    "df_info = pd.read_csv(info_src, sep = '\\t')\n",
    "print df_info.shape\n",
    "df_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'function' object has no attribute 'iloc'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-34-c6c045b09538>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mdf_info\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mdf_info\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'function' object has no attribute 'iloc'"
     ]
    }
   ],
   "source": [
    "# df_info = df_info.iloc[:,-2:].drop_duplicates().transpose()\n",
    "# df_info.columns = df_info.iloc[1]\n",
    "df_info\n",
    "\n",
    "df_info.drop(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# OK, so there are 59838 transcripts mapped.  First, how manyn are never observed?\n",
    "test = df.iloc[:,1:].sum(axis = 1)\n",
    "print 'number of genes:',test.shape[0]\n",
    "\n",
    "unobserved = 0\n",
    "for i in range(0,len(test)):\n",
    "    if test[i]==0:\n",
    "        unobserved+=1\n",
    "\n",
    "print 'number of unobserved genes:',unobserved\n",
    "print 'number of observed genes:',(test.shape[0]-unobserved)\n",
    "\n",
    "# this drops all the all zero rows:\n",
    "df_small = df.loc[~(df.select_dtypes(include=['number'])==0).all(axis=1),:]\n",
    "print df_small.shape\n",
    "df_small.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to normalize the data. \n",
    "Reads Per Million is what they mostly use in the paper.  \n",
    "RPKM (transcript length normalized) would be better to get and is also pretty standard,\n",
    "TPM does the length scaling before the per million reads scaling and might be a better metric for the number\n",
    "of transcripts.\n",
    "Transcript length normalization is a better option because of the tagmenation step in the Smart Seq2 protocol that they use.  This step breaks up transcripts to roughly equal sized chunks to improve the sequencing.  The downside is that it maps each chunk of a transcript to a unique read.\n",
    "It's paired end sequencing, so it's technically FPKM, but that's unimportant.\n",
    "Because I want to eventually look for signals that go across cells, I'm going to use TPM to normalize the number of transcripts in each cell.\n",
    "\n",
    "TPM first scales by transcript length and then by sequencing depth.\n",
    "\n",
    "So for the length... yeah, I need to get that data.  The Smart Seq 2 protocol is a poly-A selection protocol, so the length can be obtained by summing over the exons in the mature transcript.  The paper used the hg19 reference genome to map and gencode v17 for the splice junctions.  I downloaded the gtf for gencode v17 and can use that to sum over the exons and build a reference table.  That code will be in the next notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
